### PR TITLE
New modules for webapp network access restrictions

### LINKED ISSUE #379: Configure web app networking (access restrictions)
##### SUMMARY
Currently there is no way to configure networking/access restrictions on a web app.

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
azure_rm_webapp

##### ADDITIONAL INFORMATION
The desire is to be able to add allow/deny rules to a web app's networking configuration.

Additionally we should be able to configure the web app's SCM endpoint to have the same restrictions as the web app (matches portal configuration ability).

This will match azure-cli ability with `az webapp config access-restriction` (see [doc](https://docs.microsoft.com/en-us/cli/azure/webapp/config/access-restriction?view=azure-cli-latest)).

This request seems related to https://github.com/ansible-collections/azure/issues/250.

### PR DESCRIPTION
**Pull request by l3ender**
##### SUMMARY
This PR includes new modules to accomplish ability to configure a webapp's network access restrictions.

Resolves #379.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
azure_rm_webappaccessrestriction
azure_rm_webappaccessrestriction_info

##### ADDITIONAL INFORMATION

```yaml
- name: "Configure webapp network access restrictions"
  azure_rm_webappaccessrestriction:
    name: webapp{{ rpfx }}
    resource_group: "{{ resource_group }}"
    ip_security_restrictions:
      - name: "Datacenter 1"
        action: "Allow"
        ip_address: "1.2.3.4/24"
        priority: 1
      - name: "Datacenter 2"
        action: "Allow"
        ip_address: "5.6.7.8/24"
        priority: 2
    scm_ip_security_restrictions:
      - name: "Datacenter 1"
        action: "Allow"
        ip_address: "1.2.3.4/24"
        priority: 1
    scm_ip_security_restrictions_use_main: false
  register: output

- name: "Delete webapp network access restrictions"
  azure_rm_webappaccessrestriction:
    name: webapp{{ rpfx }}
    resource_group: "{{ resource_group }}"
    state: absent
  register: output
```

Output looks like:
```bash
ok: [localhost] => {
    "msg": {
        "changed": true,
        "failed": false,
        "ip_security_restrictions": [
            {
                "action": "Allow",
                "description": null,
                "ip_address": "1.2.3.4/24",
                "name": "Datacenter 1",
                "priority": 1
            },
            {
                "action": "Allow",
                "description": null,
                "ip_address": "5.6.7.8/24",
                "name": "Datacenter 2",
                "priority": 2
            }
        ],
        "scm_ip_security_restrictions": [
            {
                "action": "Allow",
                "description": null,
                "ip_address": "1.2.3.4/24",
                "name": "Datacenter 1",
                "priority": 1
            }
        ],
        "scm_ip_security_restrictions_use_main": false
    }
}

ok: [localhost] => {
    "msg": {
        "changed": true,
        "failed": false,
        "ip_security_restrictions": [],
        "scm_ip_security_restrictions": [],
        "scm_ip_security_restrictions_use_main": false
    }
}
```

### CODE REVIEW

**Comment by l3ender** [comment]
Hi @Fred-sun, here is another PR with new modules. Please review and advise. Thank you!

**Comment by Fred-sun** [comment]
@l3ender Please add module name "azure_rm_webapp_accessrestriction" to  pr-pipeline.yml.  Another, change name to "azure_rm_webappaccessrestriction" , keep same with other modules! Thank you very much!

File: plugins/modules/azure_rm_webapp_accessrestriction.py [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+                type: int
+                required: true
+            ip_address:
+                description:
+                    - IPv4 address (with subnet mask) of the access restriction.
+                type: str
+                required: true
+    scm_ip_security_restrictions:
+        description:
+            - The web app's SCM access restrictions. If C(scm_ip_security_restrictions_use_main) is set to C(true), the SCM restrictions will be configured but not used.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction.py]: line too long (169 > 160 characters)

File: plugins/modules/azure_rm_webapp_accessrestriction.py [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+                type: int
+                required: true
+            ip_address:
+                description:
+                    - IPv4 address (with subnet mask) of the access restriction.
+                type: str
+                required: true
+    scm_ip_security_restrictions_use_main:
+        description:
+            - Set to C(true) to have the HTTP access restrictions also apply to the SCM site. If C(scm_ip_security_restrictions) are also applied, they will configured but not used.
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction.py]: line too long (181 > 160 characters)

File: plugins/modules/azure_rm_webapp_accessrestriction.py [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+    pass
+
+ip_restriction_spec = dict(
+    name=dict(type='str'),
+    description=dict(type='str'),
+    action=dict(type='str', default='Allow', choices=['Allow', 'Deny']),
+    priority=dict(type='int', required=True),
+    ip_address=dict(type='str', required=True),
+)
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction.py]:
```suggestion


```

File: plugins/modules/azure_rm_webapp_accessrestriction.py [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+                self.results.update(self.set_results(site_config))
+
+        self.results['changed'] = changed
+
+        return self.results
+
+    def get_updated_config(self, site_config):
+        site_config.ip_security_restrictions = [] if not self.ip_security_restrictions else self.to_restriction_obj_list(self.ip_security_restrictions)
+        site_config.scm_ip_security_restrictions = [] if not self.scm_ip_security_restrictions else self.to_restriction_obj_list(self.scm_ip_security_restrictions)
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction.py]: line too long (163 > 160 characters)

File: plugins/modules/azure_rm_webapp_accessrestriction.py [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+        site_config.scm_ip_security_restrictions_use_main = self.scm_ip_security_restrictions_use_main
+        return site_config
+
+    def has_updates(self, site_config):
+        return (self.ip_security_restrictions != self.to_restriction_dict_list(site_config.ip_security_restrictions)
+            or self.scm_ip_security_restrictions != self.to_restriction_dict_list(site_config.scm_ip_security_restrictions)
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction.py]: continuation line under-indented for visual indent

File: plugins/modules/azure_rm_webapp_accessrestriction.py [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+            or site_config.scm_ip_security_restrictions_use_main != self.scm_ip_security_restrictions_use_main)
+
+    def has_access_restrictions(self, site_config):
+        return site_config.ip_security_restrictions or site_config.scm_ip_security_restrictions
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction.py]: continuation line under-indented for visual indent

File: plugins/modules/azure_rm_webapp_accessrestriction.py [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+    def to_restriction_dict(self, restriction_obj):
+        return dict(
+            name=restriction_obj.name,
+            description=restriction_obj.description,
+            action=restriction_obj.action,
+            priority=restriction_obj.priority,
+            ip_address=restriction_obj.ip_address,
+        )
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction.py]: expected 2 blank lines, found 1
```suggestion


```

File: plugins/modules/azure_rm_webapp_accessrestriction.py [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+def main():
+    AzureRMWebAppAccessRestriction()
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction.py]: expected 2 blank lines after class or function definition, found 1
```suggestion


```

File: plugins/modules/azure_rm_webapp_accessrestriction_info.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+            type: str
+            sample: 1.1.1.1/32
+scm_ip_security_restrictions_use_main:
+    description:
+        - Whether the HTTP access restrictions are used for SCM access.
+    returned: always
+    type: bool
+    sample: false
+'''
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction_info.py]: expected 2 blank lines, found 1

```suggestion


```

File: plugins/modules/azure_rm_webapp_accessrestriction_info.py [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+    def to_restriction_dict(self, restriction_obj):
+        return dict(
+            name=restriction_obj.name,
+            description=restriction_obj.description,
+            action=restriction_obj.action,
+            priority=restriction_obj.priority,
+            ip_address=restriction_obj.ip_address,
+        )
+
+def main():
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction_info.py]: expected 2 blank lines, found 1

```suggestion

def main():
```

File: plugins/modules/azure_rm_webapp_accessrestriction_info.py [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+    AzureRMWebAppAccessRestrictionInfo()
+
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction_info.py]: expected 2 blank lines, found 1

```suggestion


```

File: plugins/modules/azure_rm_webapp_accessrestriction.py [hunk 12]
```diff
@@ -0,0 +1,364 @@
...
+            - State of the access restrictions. Use C(present) to create or update and C(absent) to delete.
+        type: str
+        default: present
+        choices:
+            - absent
+            - present
+    ip_security_restrictions:
+        description:
+            - The web app's HTTP access restrictions.
+        type: list
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction.py]: Argument 'ip_security_restrictions' in argument_spec specifies elements as dict,but elements is not documented
```suggestion
        type: list
        elements: dict
```

File: plugins/modules/azure_rm_webapp_accessrestriction.py [hunk 13]
```diff
@@ -0,0 +1,364 @@
...
+                required: true
+            ip_address:
+                description:
+                    - IPv4 address (with subnet mask) of the access restriction.
+                type: str
+                required: true
+    scm_ip_security_restrictions:
+        description:
+            - The web app's SCM access restrictions. If C(scm_ip_security_restrictions_use_main) is set to C(true), the SCM restrictions will be configured but not used.
+        type: list
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction.py]: Argument 'scm_ip_security_restrictions' in argument_spec specifies elements as dict,but elements is not documented
```suggestion
        type: list
        elements: dict
```

File: plugins/modules/azure_rm_webapp_accessrestriction.py [hunk 14]
```diff
@@ -0,0 +1,364 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2021 Ross Bender (@l3ender)
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+from inspect import signature
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction.py]: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN

File: plugins/modules/azure_rm_webapp_accessrestriction.py [hunk 15]
```diff
@@ -0,0 +1,364 @@
...
+from ansible_collections.azure.azcollection.plugins.module_utils.azure_rm_common import AzureRMModuleBase
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction.py]: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN

File: plugins/modules/azure_rm_webapp_accessrestriction_info.py [hunk 16]
```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2021 Ross Bender (@l3ender)
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+from ansible_collections.azure.azcollection.plugins.module_utils.azure_rm_common import AzureRMModuleBase
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webapp_accessrestriction_info.py]: Import found before documentation variables. All imports must appear below DOCUMENTATION/EXAMPLES/RETURN

**Comment by Fred-sun** [comment]
@l3ender Please add use cases related to the azure_rm_webapp_accessrestriction_info.py module in tests/integration/targets/azure_rm_webapp_accessrestriction/tasks/main.yml, thank you!

**Comment by Fred-sun** [comment]
@l3ender You are advised to change the module name to --azure_rm_webappaccessrestriction. to ensure that the module name is consistent with other modules. Thank you very much!

**Comment by l3ender** [comment]
@Fred-sun PR is updated, please check again. Thank you!

File: plugins/modules/azure_rm_webappaccessrestriction.py [hunk 17]
```diff
@@ -0,0 +1,377 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = '''
+---
+module: azure_rm_webappaccessrestriction
+version_added: "1.9.0"
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webappaccessrestriction.py]:
```suggestion
version_added: "1.8.0"
```

File: plugins/modules/azure_rm_webappaccessrestriction_info.py [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+DOCUMENTATION = '''
+---
+module: azure_rm_webappaccessrestriction_info
+version_added: "1.9.0"
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webappaccessrestriction_info.py]:
```suggestion
version_added: "1.9.0"
```

File: plugins/modules/azure_rm_webappaccessrestriction_info.py [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+            ip_security_restrictions=[],
+            scm_ip_security_restrictions=[],
+            scm_ip_security_restrictions_use_main=False,
+        )
+
+        self.name = None
+        self.resource_group = None
+
+        super(AzureRMWebAppAccessRestrictionInfo, self).__init__(self.module_arg_spec,
+                                                                 supports_tags=False,
```
Comment (Fred-sun) [inline plugins/modules/azure_rm_webappaccessrestriction_info.py:151]:
```suggestion
                                                                 supports_check_mode=True,
                                                                 supports_tags=False,
```

**Review by Fred-sun — COMMENTED** [review]
@l3ender ansible core new request! Thank you very much!

**Comment by xuzhang3** [comment]
LGTM

**Pull request merged**